File: backend/models/character.py
"""
Character model - Player characters and NPCs.
"""

from sqlalchemy import Column, String, Boolean, DateTime, Integer, Text, ForeignKey, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from database import Base
from db_types import GUID, FlexJSON
import uuid
import enum


class CharacterType(str, enum.Enum):
    """Character type"""
    PLAYER = "player"
    NPC = "npc"
    AI_PLAYER = "ai_player"


class Character(Base):
    """Character model for PCs, NPCs, and AI players"""
    __tablename__ = "characters"
    
    # Primary
    id = Column(GUID(), primary_key=True, default=uuid.uuid4)
    name = Column(String(100), nullable=False)
    
    # Ownership
    user_id = Column(GUID(), ForeignKey("users.id"), nullable=True)  # Null for NPCs
    campaign_id = Column(GUID(), ForeignKey("campaigns.id"), nullable=False)
    
    # Type
    character_type = Column(
        Enum(CharacterType),
        default=CharacterType.PLAYER,
        nullable=False
    )
    is_active = Column(Boolean, default=True)
    
    # Basic info
    race = Column(String(50))
    character_class = Column(String(100))  # "class" is reserved keyword
    subclass = Column(String(100))
    level = Column(Integer, default=1)
    multiclass = Column(FlexJSON, default=[])  # [{"class": "Fighter", "level": 3}]
    background = Column(String(100))
    alignment = Column(String(50))
    
    # Appearance
    age = Column(Integer)
    height = Column(String(20))
    weight = Column(String(20))
    eyes = Column(String(50))
    skin = Column(String(50))
    hair = Column(String(50))
    
    # Stats (stored as JSON for flexibility across rule systems)
    ability_scores = Column(FlexJSON, default={})  # {"str": 10, "dex": 14, ...}
    ability_score_improvements = Column(FlexJSON, default={})  # Track ASI history
    skills = Column(FlexJSON, default={})
    saving_throws = Column(FlexJSON, default={})
    proficiency_bonus = Column(Integer, default=2)
    
    # Combat stats
    max_hp = Column(Integer, default=10)
    current_hp = Column(Integer, default=10)
    temp_hp = Column(Integer, default=0)
    armor_class = Column(Integer, default=10)
    initiative_bonus = Column(Integer, default=0)
    speed = Column(Integer, default=30)
    
    # Death saves
    death_save_successes = Column(Integer, default=0)
    death_save_failures = Column(Integer, default=0)
    
    # Hit dice
    hit_dice_total = Column(Integer, default=1)
    hit_dice_remaining = Column(Integer, default=1)
    hit_die_type = Column(String(10), default="d8")
    
    # Passive scores
    passive_perception = Column(Integer, default=10)
    passive_investigation = Column(Integer, default=10)
    passive_insight = Column(Integer, default=10)
    
    # Features
    proficiencies = Column(FlexJSON, default=[])
    languages = Column(FlexJSON, default=[])
    features = Column(FlexJSON, default=[])
    
    # Equipment (detailed tracking)
    weapons = Column(FlexJSON, default=[])
    armor = Column(FlexJSON, default={})
    inventory = Column(FlexJSON, default=[])
    equipment = Column(FlexJSON, default=[])  # Legacy field, keeping for compatibility
    attunement_slots_used = Column(Integer, default=0)
    
    # Currency
    currency = Column(FlexJSON, default={"cp": 0, "sp": 0, "ep": 0, "gp": 0, "pp": 0})
    
    # Spellcasting
    spells = Column(FlexJSON, default=[])
    spell_slots = Column(FlexJSON, default={})  # {"1": 4, "2": 3, ...}
    spell_slots_used = Column(FlexJSON, default={})
    spellcasting_ability = Column(String(3))  # INT, WIS, CHA
    spell_save_dc = Column(Integer)
    spell_attack_bonus = Column(Integer)
    
    # Class resources (Ki, Rage, Sorcery Points, etc.)
    class_resources = Column(FlexJSON, default={})
    
    # Description
    description = Column(Text)
    backstory = Column(Text)
    personality_traits = Column(Text)
    ideals = Column(Text)
    bonds = Column(Text)
    flaws = Column(Text)
    
    # Visuals
    avatar_url = Column(String(500))
    token_url = Column(String(500))  # Battle map token
    
    # AI-specific (for AI players)
    ai_personality = Column(String(50))  # cautious, brave, witty, etc.
    ai_behavior_prompt = Column(Text)
    
    # Experience
    experience_points = Column(Integer, default=0)
    
    # Status effects
    inspiration = Column(Boolean, default=False)
    exhaustion_level = Column(Integer, default=0)
    conditions = Column(FlexJSON, default=[])  # ["poisoned", "charmed"]
    
    # Notes
    dm_notes = Column(Text)  # Private DM notes
    player_notes = Column(Text)  # Player's own notes
    
    # Relationships
    active_effects = relationship("CharacterEffect", back_populates="character", cascade="all, delete-orphan")
    character_spells = relationship("CharacterSpell", back_populates="character", cascade="all, delete-orphan")
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    
    def __repr__(self):
        return f"<Character {self.name} (Lv{self.level} {self.character_class})>"
    
    @property
    def is_player_character(self) -> bool:
        return self.character_type == CharacterType.PLAYER
    
    @property
    def is_npc(self) -> bool:
        return self.character_type == CharacterType.NPC
    
    @property
    def is_ai_controlled(self) -> bool:
        return self.character_type == CharacterType.AI_PLAYER
    
    @property
    def is_alive(self) -> bool:
        return self.current_hp > 0
    
    @property
    def is_bloodied(self) -> bool:
        """Character is below 50% HP"""
        return self.current_hp <= (self.max_hp / 2)

File: frontend/lib/types/character.ts
// Character types matching backend models

export type CharacterType = 'player' | 'npc' | 'ai_player'

export interface AbilityScores {
  str: number
  dex: number
  con: number
  int: number
  wis: number
  cha: number
}

export interface Skill {
  name: string
  ability: keyof AbilityScores
  proficient: boolean
  expertise?: boolean
}

export interface SavingThrow {
  ability: keyof AbilityScores
  proficient: boolean
}

export interface Equipment {
  id: string
  name: string
  type: string
  quantity: number
  weight: number
  equipped?: boolean
  description?: string
  properties?: string[]
  damage?: string
  ac?: number
}

export interface Spell {
  id: string
  name: string
  level: number
  school: string
  castingTime: string
  range: string
  components: string[]
  duration: string
  description: string
  prepared?: boolean
  ritual?: boolean
  concentration?: boolean
}

export interface SpellSlots {
  [level: number]: {
    max: number
    current: number
  }
}

export interface Feature {
  name: string
  source: string  // Race, Class, Background, etc.
  description: string
}

export interface Character {
  id: string
  name: string
  user_id?: string
  campaign_id: string
  character_type: CharacterType
  is_active: boolean
  
  // Basic info
  race: string
  character_class: string
  subclass?: string
  level: number
  multiclass?: Array<{ class: string; level: number }>
  background: string
  alignment: string
  
  // Appearance
  age?: number
  height?: string
  weight?: string
  eyes?: string
  skin?: string
  hair?: string
  
  // Stats
  ability_scores: AbilityScores
  ability_score_improvements?: Record<string, any>
  proficiency_bonus: number
  skills: { [key: string]: Skill }
  saving_throws: { [key: string]: SavingThrow }
  passive_perception: number
  passive_investigation: number
  passive_insight: number
  
  // Combat
  max_hp: number
  current_hp: number
  temp_hp: number
  armor_class: number
  initiative_bonus: number
  speed: number
  
  // Hit dice (legacy and new)
  hit_dice?: {
    total: number
    current: number
    type: string  // e.g., "d8"
  }
  hit_dice_total?: number
  hit_dice_remaining?: number
  hit_die_type?: string
  
  // Death saves (legacy and new)
  death_saves?: {
    successes: number
    failures: number
  }
  death_save_successes?: number
  death_save_failures?: number
  
  // Features & abilities
  proficiencies: string[]
  languages: string[]
  features: Feature[]
  equipment: Equipment[]
  
  // Detailed equipment tracking
  weapons?: Array<{
    name: string
    damage: string
    damage_type: string
    properties?: string[]
    equipped?: boolean
  }>
  armor?: {
    name: string
    ac: number
    type: string
    equipped?: boolean
  }
  inventory?: Array<{
    name: string
    quantity: number
    weight?: number
    description?: string
  }>
  attunement_slots_used?: number
  
  // Currency
  currency?: {
    cp: number
    sp: number
    ep: number
    gp: number
    pp: number
  }
  
  // Spellcasting
  spells: Spell[]
  spell_slots?: SpellSlots
  spell_slots_used?: Record<number, number>
  spellcasting_ability?: 'INT' | 'WIS' | 'CHA'
  spell_save_dc?: number
  spell_attack_bonus?: number
  
  // Class resources (Ki, Rage, Sorcery Points, etc.)
  class_resources?: Record<string, {
    max: number
    current: number
    name: string
  }>
  
  // Status effects
  inspiration?: boolean
  exhaustion_level?: number
  conditions?: string[]
  
  // Description
  description?: string
  backstory?: string
  personality_traits?: string
  ideals?: string
  bonds?: string
  flaws?: string
  
  // Visuals
  avatar_url?: string
  token_url?: string
  
  // AI-specific
  ai_personality?: string
  ai_behavior_prompt?: string
  
  // Progress
  experience_points: number
  
  // Notes
  dm_notes?: string
  player_notes?: string
  
  // Timestamps
  created_at: string
  updated_at?: string
}

export interface CharacterCreateData {
  name: string
  campaign_id: string
  character_type?: CharacterType
  race: string
  character_class: string
  level?: number
  background: string
  alignment: string
  ability_scores: AbilityScores
  max_hp?: number
  armor_class?: number
  speed?: number
  description?: string
  backstory?: string
  personality_traits?: string
  ideals?: string
  bonds?: string
  flaws?: string
  dm_notes?: string
  avatar_url?: string
}

export interface CharacterUpdateData {
  name?: string
  level?: number
  current_hp?: number
  temp_hp?: number
  max_hp?: number
  armor_class?: number
  ability_scores?: AbilityScores
  skills?: { [key: string]: Skill }
  equipment?: Equipment[]
  spells?: Spell[]
  spell_slots?: SpellSlots
  experience_points?: number
  description?: string
  backstory?: string
  personality_traits?: string
  ideals?: string
  bonds?: string
  flaws?: string
  player_notes?: string
  avatar_url?: string
}

// Helper functions
export function calculateModifier(score: number): number {
  return Math.floor((score - 10) / 2)
}

export function calculateProficiencyBonus(level: number): number {
  return Math.ceil(level / 4) + 1
}

export function getSkillModifier(
  character: Character,
  skillName: string
): number {
  const skill = character.skills[skillName]
  if (!skill) return 0
  
  const abilityScore = character.ability_scores[skill.ability]
  const abilityMod = calculateModifier(abilityScore)
  const profBonus = calculateProficiencyBonus(character.level)
  
  let modifier = abilityMod
  if (skill.proficient) {
    modifier += profBonus
  }
  if (skill.expertise) {
    modifier += profBonus // Double proficiency for expertise
  }
  
  return modifier
}

export function getSavingThrowModifier(
  character: Character,
  ability: keyof AbilityScores
): number {
  const abilityScore = character.ability_scores[ability]
  const abilityMod = calculateModifier(abilityScore)
  const profBonus = calculateProficiencyBonus(character.level)
  
  const savingThrow = character.saving_throws[ability]
  const isProficient = savingThrow?.proficient || false
  
  return abilityMod + (isProficient ? profBonus : 0)
}

export function formatModifier(modifier: number): string {
  return modifier >= 0 ? `+${modifier}` : `${modifier}`
}

// D&D 5e standard skills
export const STANDARD_SKILLS: { name: string; ability: keyof AbilityScores }[] = [
  { name: 'Acrobatics', ability: 'dex' },
  { name: 'Animal Handling', ability: 'wis' },
  { name: 'Arcana', ability: 'int' },
  { name: 'Athletics', ability: 'str' },
  { name: 'Deception', ability: 'cha' },
  { name: 'History', ability: 'int' },
  { name: 'Insight', ability: 'wis' },
  { name: 'Intimidation', ability: 'cha' },
  { name: 'Investigation', ability: 'int' },
  { name: 'Medicine', ability: 'wis' },
  { name: 'Nature', ability: 'int' },
  { name: 'Perception', ability: 'wis' },
  { name: 'Performance', ability: 'cha' },
  { name: 'Persuasion', ability: 'cha' },
  { name: 'Religion', ability: 'int' },
  { name: 'Sleight of Hand', ability: 'dex' },
  { name: 'Stealth', ability: 'dex' },
  { name: 'Survival', ability: 'wis' },
]

export const ABILITY_NAMES: { [key in keyof AbilityScores]: string } = {
  str: 'Strength',
  dex: 'Dexterity',
  con: 'Constitution',
  int: 'Intelligence',
  wis: 'Wisdom',
  cha: 'Charisma',
}
